referenceframe/frame_system.go

```diff
@@ -79,12 +80,77 @@ type simpleFrameSystem struct {
 	parents map[Frame]Frame
 }
 
-// NewEmptySimpleFrameSystem creates a graph of Frames that have.
-func NewEmptySimpleFrameSystem(name string) FrameSystem {
+// NewEmptyFrameSystem creates a graph of Frames that have.
+func NewEmptyFrameSystem(name string) FrameSystem {
 	worldFrame := NewZeroStaticFrame(World)
 	return &simpleFrameSystem{name, worldFrame, map[string]Frame{}, map[Frame]Frame{}}
 }
 
+// NewFrameSystem assembles a frame system from a set of parts and additional transforms.
+func NewFrameSystem(name string, parts []*FrameSystemPart, additionalTransforms []*LinkInFrame) (FrameSystem, error) {
+	allParts := make([]*FrameSystemPart, 0)
+	allParts = append(allParts, parts...)
+	for _, tf := range additionalTransforms {
+		transformPart, err := LinkInFrameToFrameSystemPart(tf)
+		if err != nil {
+			return nil, err
+		}
+		allParts = append(allParts, transformPart)
+	}
+
+	// ensure that at least one frame connects to world if the frame system is not empty
+	if len(allParts) != 0 {
+		hasWorld := false
+		for _, part := range allParts {
+			if part.FrameConfig.Parent() == World {
+				hasWorld = true
+				break
+			}
+		}
+		if !hasWorld {
+			return nil, ErrNoWorldConnection
+		}
+	}
+	// Topologically sort parts
+	sortedParts, err := TopologicallySortParts(allParts)
```

Comment:
Its currently still being used in the framesystem service and I expect it will continue to be, since we can cache the list of parts after it has been sorted to reduce extra operations down the line.